### PR TITLE
Produce junit results for verify job

### PR DESCRIPTION
**Pull request by cblecker**
**What this PR does / why we need it**:
Junit results may make the verify job clearer, and give us better insights into timings

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by cblecker** [comment]
@bentheelder I’m guess there might also be some changes to the dind job to collect junit artifacts? I’m seeing the file locally, but not in gubernator.

**Comment by BenTheElder** [comment]
@cblecker the DIND job shouldn't be doing anything special, bootstrap just scoops up anything in artifacts and gubernator reads these. It looks like none were uploaded http://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/58783/pull-kubernetes-verify/73591/artifacts/

**Comment by cblecker** [comment]
@BenTheElder Found it. Needed to set this in the "jenkins" script: https://github.com/kubernetes/kubernetes/pull/58783/commits/4986d4f92a4c617c3f41632c14b8d119df813a0a

File: hack/make-rules/verify.sh
Comment (fejta) [inline hack/make-rules/verify.sh]: `>&2`?

**Comment by fejta** [comment]
Feel free to unhold when/if you're satisfied with the feedback from ben or jeff

/cc @BenTheElder

**Review by BenTheElder — APPROVED** [review]
/lgtm

**Comment by BenTheElder** [comment]
@ixdy @stevekuznetsov perhaps you two could check for any bash quirks?
Thanks @cblecker, having individual results for these will be great!

**Comment by cblecker** [comment]
Well that's looking nice: https://k8s-testgrid.appspot.com/presubmits-kubernetes-blocking#pull-kubernetes-verify

Let's see how it looks when I break things.

**Comment by cblecker** [comment]
Well there's problem number 1 for tomorrow:
```
I0125 06:14:26.035] Run ./hack/update-bazel.sh
I0125 06:14:26.859] +++ exit code: 1
I0125 06:14:26.863] +++ error: 1
I0125 06:14:26.924] SUCCESS  verify-bazel.sh	19s
```
It's erroring, but the error code isn't being passed on to the main script. That won't work

**Comment by fejta** [comment]
@stevekuznetsov did some bash hackery here using `trap`: https://github.com/kubernetes/test-infra/blob/master/hack/update-deps.sh#L49

File: third_party/forked/shell2junit/sh2ju.sh
Comment (stevekuznetsov) [inline third_party/forked/shell2junit/sh2ju.sh]: This seems like a big ask for every dev on Mac -- are we ready to do that? Did we ask mailing lists?

File: hack/make-rules/verify.sh
Comment (stevekuznetsov) [inline hack/make-rules/verify.sh]: This is smelly

File: hack/make-rules/verify.sh
Comment (stevekuznetsov) [inline hack/make-rules/verify.sh]: When would this be true? Isn't it checked in to the repo as well?

File: cmd/hyperkube/hyperkube.go
Comment (stevekuznetsov) [inline cmd/hyperkube/hyperkube.go]: Seems like a typo?

File: hack/make-rules/verify.sh
Comment (stevekuznetsov) [inline hack/make-rules/verify.sh]: Always prefer to be explicit

```sh
if [[ -n "${SILENT:-}" ]]; then
```

Or even better, check for the specific string you expect to be passed

**Comment by stevekuznetsov** [comment]
Looks like maybe your `juLog` isn't propagating it's error code?

**Comment by stevekuznetsov** [comment]
FYI if you want less Bash in your life I also wrote https://github.com/stevekuznetsov/exec-assert which may do what you need :)

**Comment by stevekuznetsov** [comment]
Actually, I never published the jUnit part of that. Nevermind.

File: third_party/forked/shell2junit/sh2ju.sh
Comment (cblecker) [reply third_party/forked/shell2junit/sh2ju.sh]: We already do this in a number of scripts using this function: https://github.com/kubernetes/kubernetes/blob/884e4b7b9c63a9a3b94278972d640454b8d16e47/hack/lib/util.sh#L779-L794

I've just copy pasted it over into the forked junit code, because I was getting errors with the SEDs in it on my Mac.

File: hack/make-rules/verify.sh
Comment (cblecker) [reply hack/make-rules/verify.sh]: It is, but seems to be a requirement for the forked shell2junit code. We already do it here: https://github.com/kubernetes/kubernetes/blob/884e4b7b9c63a9a3b94278972d640454b8d16e47/hack/make-rules/test-cmd-util.sh#L91-L114

File: hack/make-rules/verify.sh
Comment (cblecker) [reply hack/make-rules/verify.sh]: It is. I don't know why this wouldn't trigger, but I had copied it from here: https://github.com/kubernetes/kubernetes/blob/884e4b7b9c63a9a3b94278972d640454b8d16e47/hack/make-rules/test-cmd-util.sh#L82-L89

File: cmd/hyperkube/hyperkube.go
Comment (cblecker) [reply cmd/hyperkube/hyperkube.go]: It is. This is part of my commit that breaks things. :)

**Comment by cblecker** [comment]
See, now that's not too bad: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/58783/pull-kubernetes-verify/73694/

File: third_party/forked/shell2junit/sh2ju.sh
Comment (stevekuznetsov) [reply third_party/forked/shell2junit/sh2ju.sh]: ACK

File: hack/make-rules/verify.sh
Comment (stevekuznetsov) [reply hack/make-rules/verify.sh]: Maybe we shouldn't fork a lib that isn't hygenic? :)

File: hack/make-rules/verify.sh
Comment (stevekuznetsov) [reply hack/make-rules/verify.sh]: Maybe we can drop it then, seems like a bit of a cargo cult situation :D

**Comment by cblecker** [comment]
Fixed up the unset stuff in the forked script. PTAL @stevekuznetsov

File: hack/make-rules/verify.sh
Comment (stevekuznetsov) [inline hack/make-rules/verify.sh]: If we're running with `set -o errexit` on why does this work?

**Comment by fejta** [comment]
/approve

Steve want to toss on an LGTM when you like how it looks? Your bash skills are beyond mine

**Comment by fejta** [comment]
/assign @ixdy 
For the shell2junit change

File: hack/make-rules/verify.sh
Comment (ixdy) [reply hack/make-rules/verify.sh]: we capture the status when calling `run-cmd`, which is why the script doesn't fully exit on failures here.

I don't think we need to do this `tr=$?` and `return ${tr}` stuff, though - we can keep it as it was before; the correct status should be returned regardless.

**Comment by ixdy** [comment]
minor nit, but I don't think that should block this merging - those are much nicer results on gubernator.

/approve

File: hack/make-rules/verify.sh
Comment (cblecker) [reply hack/make-rules/verify.sh]: This is likely the reason: https://stackoverflow.com/a/37191242/

Comment (cblecker) [reply hack/make-rules/verify.sh]: @ixdy It [wasn't returning the status](https://github.com/kubernetes/kubernetes/pull/58783#issuecomment-360375698), so these were added.

Comment (ixdy) [reply hack/make-rules/verify.sh]: wait, how did this ever work?

Comment (cblecker) [reply hack/make-rules/verify.sh]: ✨bash✨

**Review by stevekuznetsov — COMMENTED** [review]
Two comments that can be done in a follow-up, I guess.

```sh
run-cmd "${runner}" "${t}" && tr=$? || tr=$?
```

This is a massive anti-pattern for two reasons:

```sh
#!/bin/bash

set -o errexit
set -o pipefail
set -o nounset

function control-flow() {
	echo "should run"
	false
	echo "should never run"
}

control-flow || r=$?

true && false || echo "should not echo"
```

```
$ bad.sh
should run
should never run
should not echo
```

Any function inside of an `if` block of before a `||` statement has it's `ERR` pseudo-signal turned off and therefore any control flow that would have been expected from an early exit using `errexit` will not occur. This is surprising and hidden behavior that needs to be carefully managed.

Secondly, `expr && a || b` is not a ternary operator, as seen in the example. The real behavior is:

```sh
if expr; then
    if ! a; then
        b
    fi
else 
    b
fi
```

**Comment by stevekuznetsov** [comment]
The benefits here are nice, there is time to do a follow-up before someone shoots themselves in the foot with Bash later :)

/lgtm

**Comment by cblecker** [comment]
Found some more unbound variable fun. Opened https://github.com/kubernetes/kubernetes/pull/58889

**Comment by CaoShuFeng** [comment]
Ref:
https://github.com/kubernetes/kubernetes/pull/58907

**Pull request merged**
